### PR TITLE
[6X] Do not preserve relfrozenxid of user tables for pg_dump --binary-upgrade

### PR DESCRIPTION
**Pull request by jimmyyih**
When running pg_dump with --binary-upgrade, the relfrozenxids are preserved for each user table. However, there were many issues with the logic:
1. For partition tables, there was no preservation SQL commands outputted for subroot and child partitions. Only the root partition had a SQL command be outputted to preserve its relfrozenxid.
2. For AO/CO tables, the relfrozenxid does not need to be preserved since AO/CO tuples do not contain xid info.
3. After pg_upgrade runs pg_restore to recreate objects and update relfrozenxids as per the dump file, pg_upgrade later bulk updates pg_class to set the relfrozenxids to the source segment's datfrozenxid for applicable pg_class tuples... which seemingly makes the relfrozenxid preservation SQL commands in the dump file redundant.

Because of the above reasons, we disable the preservation of relfrozenxid for user tables. This should help a bit in speeding up upgrade and reducing the dump file size.

Note: This change is not going into main (7X) because its relfrozenxid handling for upgrade is currently different.  If we align the logic later when we work on 6-to-7 upgrade, this patch may or may not be useful to forward-port.

gpupgrade dev pipeline: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:jimmyyih-main

### CODE REVIEW

**Comment by huansong** [comment]
> After pg_upgrade runs pg_restore to recreate objects and update relfrozenxids as per the dump file, pg_upgrade later bulk updates pg_class to set the relfrozenxids to the source segment's datfrozenxid for applicable pg_class tuples... which seemingly makes the relfrozenxid preservation SQL commands in the dump file redundant.

Interesting.. if that's the case it would make the relfrozenxids preservation pointless I agree. However, at the point where pg_upgrade sets relfrozenxids (`set_frozenxids`), the dump script hasn't been invoked right? According to the comment for `set_frozenxids`:
```
/*
 *      set_frozenxids()
 *
 * This is called on the new cluster before we restore anything, with
 * minmxid_only = false.
```
Does it mean the preserved relfrozenxids are still useful?

I agree with the partitioned table and AO table aspects though (just not sure if they are enough to remove the preservation logic as a whole).

**Review by kalensk — APPROVED** [review]

**Comment by jimmyyih** [comment]
> However, at the point where pg_upgrade sets relfrozenxids (`set_frozenxids`), the dump script hasn't been invoked right? According to the comment for `set_frozenxids`:
>
> ```
> /*
> *      set_frozenxids()
> *
> * This is called on the new cluster before we restore anything, with
> * minmxid_only = false.
> ```
>
> Does it mean the preserved relfrozenxids are still useful?

In regards to relfrozenxids, pg_upgrade is a mess of multiple refactors that seem to either contradict or compliment each other.  We have `set_frozenxids`, `update_db_xids`, `freeze_master_data`, and the restore script's relfrozenxid preservation UPDATE commands that all do similar things in relation to datfrozenxid and relfrozenxid.  The call stack of main -> prepare_new_databases -> set_frozenxids(false) is indeed done before the restore script is invoked to make sure the fresh new catalog tables are set up correctly before calling pg_restore (or else the CREATE TABLE commands would likely fail)... but that's just the one call.  Then during pg_restore, the preservation UPDATE commands are executed which kinda become redundant since we later call `update_db_xids` and `freeze_master_data` to again do relfrozenxid updates.

It's pretty tangled and can be hard to follow... but disabling the restore script's preservation logic is one step towards untangling the mess since it seems like it wasn't doing a lot of things correctly anyways and was hidden by the other relfrozenxid logics.

**Comment by jimmyyih** [comment]
I've updated some comments relating to the restore script that may help with understanding the logic flow for later work.

**Comment by huansong** [comment]
> during pg_restore, the preservation UPDATE commands for the user tables are executed which kinda become redundant since we later call update_db_xids and freeze_master_data to again do relfrozenxid updates.

I see, makes sense!

> I've updated some comments relating to the restore script that may help with understanding the logic flow for later work.

Thanks for adding it!

**Review by huansong — APPROVED** [review]
LGTM
